src/t8_forest/t8_forest_pfc_message.hxx

```diff
@@ -0,0 +1,314 @@
+   * \param[in]     buf_size  the size of the buffer
+   * \param[in,out] position  current position within the receive buffer
+   *                            on input:  0
+   *                            on output: buf_size
+   */
+  void
+  unpack (void *buf, int buf_size, int *position)
+  {
+    /* unpack: itree */
+    sc_MPI_Unpack (buf, buf_size, position, &itree, 1, T8_MPI_GLOIDX, comm);
+
+    /* unpack: eclass */
+    int eclass_int;
+    sc_MPI_Unpack (buf, buf_size, position, &eclass_int, 1, sc_MPI_INT, comm);
+    eclass = (t8_eclass_t) eclass_int;
+
+    /* unpack: parent */
+    t8_element_new (myscheme, eclass, 1, &parent);
+    allocated_parent = true;
+    t8_element_MPI_Unpack (myscheme, eclass, buf, buf_size, position, &parent, 1, comm);
+
+    /* unpack: num_siblings */
+    sc_MPI_Unpack (buf, buf_size, position, &num_siblings, 1, sc_MPI_INT, comm);
+  }
+
+  /**
+   * Determine and return the pack size of the PFC message.
+   *
+   * It results from the data fields to send and the associated sizes:
+   *  + itree (T8_MPI_GLOIDX)
+   *  + eclass (sc_MPI_INT)
+   *  + parent (via t8_element_MPI_Pack_size)
+   *  + num_siblings (sc_MPI_INT)
+   *
+   * \return The pack size of the message.
+  */
+  int
+  pack_size ()
+  {
+    /* initialize sum*/
+    int message_size = 0;
+    int datasize;
+
+    /* add size: itree */
+    sc_MPI_Pack_size (1, T8_MPI_GLOIDX, comm, &datasize);
+    message_size += datasize;
+
+    /* add size: eclass */
+    sc_MPI_Pack_size (1, sc_MPI_INT, comm, &datasize);
+    message_size += datasize;
+
+    /* add size: parent */
+    if (parent != NULL) {
+      // t8_eclass_scheme_c *scheme = schemes->eclass_schemes[eclass];
+      // scheme->t8_element_MPI_Pack_size (1, comm, &datasize);
+      t8_element_MPI_Pack_size (myscheme, eclass, 1, comm, &datasize);
+      message_size += datasize;
+    }
+
+    /* add size: num_siblings */
+    sc_MPI_Pack_size (1, sc_MPI_INT, comm, &datasize);
+    message_size += datasize;
+    return message_size;
+  }
+
+  /**
+   * Send the message using non-blocking MPI send.
+   *
+   * \param[in]   forest  the forest (only used for forest->mpicomm)
+   * \param[out]  request the MPI send request
+  */
+  void
+  mpi_Isend (t8_forest_t forest, sc_MPI_Request &request)
+  {
+    /* Allocate buffer */
+    int buffer_size = pack_size ();
+    char *send_buffer = T8_ALLOC (char, buffer_size);
+
+    /* Pack message to buffer */
+    int position = 0;
+    pack (send_buffer, buffer_size, &position);
+
+    /* Send buffer */
+    int mpiret = sc_MPI_Isend (send_buffer, position, sc_MPI_PACKED, iproc, message_tag, forest->mpicomm, &request);
+    SC_CHECK_MPI (mpiret);
+    T8_FREE (send_buffer);
+  }
+
+  /**
+   * Receive data from another process.
+   *
+   * \param[out]  recv_buf  the receive buffer
+   * \param[out]  buf_size  the size of the buffer
+  */
+  void
+  mpi_Recv (char *&recv_buf, int &buf_size)
```

Comment:
Do we always have to probe before receiving the message? Or do we know the size of the receiving message in advance?